### PR TITLE
Security fix: restrict application message handling to Bridge contract

### PR DESCRIPTION
**Pull request by denalimarsh**
This PR update applications with a simple `msg.sender` check so that only Bridge can call the handle method on applications. The Bridge contract registers its address on each application during initial on-contract app verification. Messages submitted to the Bridge by the relayer will be routed from Bridge contract -> application contract successfully and other actors won’t be able to send fabricated messages directly to the applications.

### CODE REVIEW

File: ethereum/contracts/ETHApp.sol [hunk 1]
```diff
@@ -20,6 +21,12 @@ contract ETHApp is Application {
         totalETH = 0;
     }
 
+    function register(address _bridge) public override returns(bool) {
+        require(bridge == address(0), "Bridge has already been registered");
```
Comment (vgeddes) [inline ethereum/contracts/ETHApp.sol:25]: Are you sure this shouldnt be `require(bridge != address(0)`?

File: ethereum/contracts/ETHApp.sol [hunk 2]
```diff
@@ -20,6 +21,12 @@ contract ETHApp is Application {
...
+        bridge = _bridge;
+        return true;
```
Comment (vgeddes) [inline ethereum/contracts/ETHApp.sol:27]: Rather just make this a `void` method if it is going to return `true` under all expected conditions.

File: ethereum/contracts/Bridge.sol [hunk 3]
```diff
@@ -61,6 +57,8 @@ contract Bridge {
         require(codehash != 0x0, "There's no account for this address on the network");
         require(codehash != NON_CONTRACT_ACCOUNT_HASH, "Only contract accounts can be registered as applications");
 
-        return true;
+        Application app = Application(_appID);
+        require(app.register(address(this)), "Application failed registration");
```
Comment (vgeddes) [inline ethereum/contracts/Bridge.sol:61]: `app.register` always returns true, so the `require` seems unnecessary.

**Review by vgeddes — APPROVED** [review]
Looks good, though added a few minor comments.

File: ethereum/contracts/Application.sol [hunk 4]
```diff
@@ -4,4 +4,6 @@ pragma solidity >=0.6.2;
 // Application contains methods that all applications must implement
 abstract contract Application {
     function handle(bytes memory _data) public virtual;
+
+    function register(address _bridge) public virtual returns(bool);
```
Comment (vgeddes) [inline ethereum/contracts/Application.sol:8]: We'll need doc comments for both `handle` and `register` so we can generate nice Solidity docs.

File: ethereum/contracts/Bridge.sol (hunk 3 above)
Comment (musnit) [reply ethereum/contracts/Bridge.sol:61]: in our current apps yes, but may be useful for future apps to allow failing registration?

File: ethereum/contracts/ETHApp.sol (hunk 2 above)
Comment (musnit) [reply ethereum/contracts/ETHApp.sol:27]: as above, think its true in the case of ETHApp but may not be true for every app

File: ethereum/contracts/Application.sol (hunk 4 above)
Comment (denalimarsh) [reply ethereum/contracts/Application.sol:14]: https://github.com/Snowfork/polkadot-ethereum/pull/115/commits/beb6cde3e8933817fca0f11e89e4af106c1bffb2

File: ethereum/contracts/Bridge.sol (hunk 3 above)
Comment (denalimarsh) [reply ethereum/contracts/Bridge.sol]: - Vincent: Already registered applications will throw an error here on `app.register(address(this))` due to the check `require(bridge == address(0), "Bridge has already been registered");` so it's redundant.
- Aidan: Future apps should fail registration if they're already registered, which is handled by the current code.

I've updated the contracts in https://github.com/Snowfork/polkadot-ethereum/pull/115/commits/ce891d0610f5a38410cd1ff327087dda266cc932

File: ethereum/contracts/ETHApp.sol (hunk 2 above)
Comment (denalimarsh) [reply ethereum/contracts/ETHApp.sol]: Went with the void + exception approach in https://github.com/Snowfork/polkadot-ethereum/pull/115/commits/ce891d0610f5a38410cd1ff327087dda266cc932

**Pull request merged**